### ISSUE TITLE
CNN experiments

### ISSUE THREAD

**Issue by danpovey** [kaldi-asr/kaldi#1569]
@GaofengCheng, I was hoping  you could help with this (this is in the kaldi_52 branch).
In wsj/s5/local/chain/tuning/run_cnn_tdnn_1a.sh I have an example of using CNNs, but it's not working great.
Something that I think is worth trying is as follows:

  - Start from our best setup, say a TDNN+LSTM setup with dropout and batch-norm.
  - Try changing the feature type (first do this without using CNNs, just with the same config as before).  I recommend using 64 mel bins and ceps instead of 40 (so change --num-mel-bins and --num-ceps both to 64).   And make sure it has --use-energy=false, otherwise it's not suitable for CNNs because it's not just a linear transformation of the filterbanks.  It looks like our hires configs already have --use-energy=false
  - Next, try adding just 2 convolutional layers at the start of those networks, with small dimensions so that it's not too slow.  And remove the learning-rate-factor=2.0 on the ivector branch, I realized it's inappropriate, and that branch is training too fast.

```
  # pre-process the ivector with fully connected layers; later we will combine                                                                              
  # it with the output of the CNN layers.                                                                                                                   
  # The learning-rate-factor >1.0 is because these layers will be instantiated                                                                              
  # once per sequence (for t==0), which will tend to make it train too slowly.                                                                              
  relu-renorm-layer name=ivector-1 input=ivector dim=200                                                                    
  relu-renorm-layer name=ivector-2 dim=200
                                                                                                                                                            
  conv-batchnorm-layer name=cnn1 input=idct height-in=64 height-out=64 time-offsets=-1,0,1 height-offsets=-1,0,1 num-filters-out=32                         
  conv-batchnorm-layer name=cnn2 height-in=64 height-out=64 time-offsets=-1,0,1 height-offsets=-1,0,1 num-filters-out=32      
 
  # the first TDNN layer:
   relu-renorm-layer input=Append(-1,0,1,ReplaceIndex(ivector-2, t, 0)) ....

```

It will be necessary to try a version of this experiment without the iVectors (you can re-use the egs by including the input node but just not using it, i.e. replacing `relu-renorm-layer input=Append(-1,0,1,ReplaceIndex(ivector-2, t, 0))...` with `relu-renorm-layer input=Append(-1,0,1)`.  The reason is that with this new way of processing ivectors, we will want to verify that they are still helping.

If this works,  we may want to have a differently named config for the feature type, e.g. hires -> xhires, and have a different version of the run_ivector_common.sh  (or different options to it).  But for now, since we don't know if it will work, you can just edit the mfcc_hires.conf and rerun the feature extraction and ivector-extractor training.

**Comment by GaofengCheng** [comment]
I will help do, the results and tuning scripts will be made into PR under kaldi_52

**Comment by danpovey** [comment]
@GaofengCheng, I just realized that I made a really stupid mistake when I implemented the xconfig layers for convolution.  I forgot to include any nonlinearity!  There is no ReLU.
I'll be modifying the xconfig code and the associated example.  It should be done before you get to the CNN stage of these experiments.

**Comment by GaofengCheng** [comment]
@danpovey I have not yet begun the CNN exps. After your modification, I'll take this and help check

**Comment by danpovey** [comment]
OK, I have fixed the CNN layer, adding ReLU's.  I'm running the
(now-modified) WSJ example and it does lead to better objective functions.

**Comment by GaofengCheng** [comment]
@danpovey get it, I'll begin CNN today

**Comment by danpovey** [comment]
BTW, something to keep an eye on is, whether the 'test-mode' batch-norm is behaving the same as the training-mode in terms of log-prob.  (we use test mode in the compute_prob logs and in the combination and in actual testing).
In the current egs/wsj/s5/local/chain/tuning/run_cnn_tdnn_1a.sh, after training a while, the training and valid log-probs become not that good, and when I tried nnet3-chain-compute-prob with --test-mode=false they got better (almost a factor of 2).  So test-mode of the BatchNorm component is not working as we expect it to.

In that configuration, though, I had converted all the layers to batch-norm, even the non-convolutional layers; and in those cases the batch-norm has to estimate many more individual scaling factors.  So it's plausible that it's somehow giving information about the current minibatch.  I also suspect some kind of bug in the batch-norm code, but I'm looking and I can't see one.

**Comment by danpovey** [comment]
I figured out the problem with the batch-norm.
It's that during a single iteration the model was moving too far, so that the batch-norm stats accumulated over the entire iteration became non-representative.
Part of the issue was that there was a bug in the reading and writing code, whereby the learning-rate and max-change was not  being read or written (and also in the xconfig I had forgotten to set the max-change to 0.75, so I had not detected that issue).
If the problem persists when the learning-rate and max-change code is fixed, I'll change how the stats for batch-norm are collected.

**Comment by GaofengCheng** [comment]
@danpovey 
How should I change cepstral-lifter if I use 64dim feature
`idct-layer name=idct input=input dim=40 cepstral-lifter=22 affine-transform-file=$dir/configs/idct.mat`

**Comment by danpovey** [comment]
leave it the same

**Comment by GaofengCheng** [comment]
@danpovey I did not find ` learning-rate-factor` related to
`And remove the learning-rate-factor=2.0 on the ivector branch, I realized it's inappropriate, and that branch is training too fast` in final.config when running cnn_tdnn_lstm or cnn_tdnn, just `learning-rate-factor=5.0 ` on the xent_output

**Comment by danpovey** [comment]
It's OK, I may already have done that.

**Comment by danpovey** [comment]
You may notice that in in egs/wsj/s5/local/chain/tuning/run_cnn_tdnn_1b.sh, in the first convolutional layer I added:
`learning-rate-factor=0.3 33 max-change=0.25`
to make it train slower, because I noticed the first convolutional layer was training too fast.  You might want to do that, or at least look out for how fast it's training (grep for Relative in progress.*.log to see).

**Comment by GaofengCheng** [comment]
yes, I'm running a comparison exp under AMI IHM to figure out whether this will help.

**Comment by GaofengCheng** [comment]
@danpovey  training setting  `   --trainer.deriv-truncate-margin 8 \` will cause the cnn_tdnn_lstm to fail (I change directly from egs/ami/s5b/tuning/run_tdnn_lstm_1i.sh). I remove this option, will this affect the results a lot?

**Comment by danpovey** [comment]
That would be a bug in the code.  I need to look into that.  Can you please
show me the error message, and if you think it would help, get me a stack
trace?

**Comment by GaofengCheng** [comment]
@danpovey 
```
nnet3-am-copy --raw=true --learning-rate=0.002 exp/ihm/chain_cleaned/tdnn_lstm1i_cnn_64_fbank_2cnn_sp_bi_ld5/0.mdl - 
LOG (nnet3-am-copy[5.1]:main():nnet3-am-copy.cc:140) Copied neural net from exp/ihm/chain_cleaned/tdnn_lstm1i_cnn_64_fbank_2cnn_sp_bi_ld5/0.mdl to raw format as -
nnet3-chain-shuffle-egs --buffer-size=5000 --srand=0 ark:- ark:- 
nnet3-chain-merge-egs --minibatch-size=32 ark:- ark:- 
nnet3-chain-copy-egs --left-context=57 --right-context=27 --frame-shift=2 ark:/nobackup/datapool/project/proa/chenggaofeng/ami/s5b_ASRU2017/exp/ihm/chain_cleaned/tdnn_lstm1i_cnn_64_fbank_2cnn_sp_bi_ld5/egs/cegs.2.ark ark:- 
ASSERTION_FAILED (nnet3-chain-train[5.1]:FindNumLeadingAndTrailingIdenticals():nnet-optimize-utils.cc:2209) : 'ptr != end && "Vector consists entirely of -1's."' 

[ Stack-Trace: ]
nnet3-chain-train() [0x11829a6]
kaldi::MessageLogger::HandleMessage(kaldi::LogMessageEnvelope const&, char const*)
kaldi::MessageLogger::~MessageLogger()
kaldi::KaldiAssertFailure_(char const*, char const*, int, char const*)
nnet3-chain-train() [0xe069c8]
nnet3-chain-train() [0xe06b0b]
kaldi::nnet3::SnipRowOps(kaldi::nnet3::NnetComputation*)
kaldi::nnet3::Optimize(kaldi::nnet3::NnetOptimizeOptions const&, kaldi::nnet3::Nnet const&, int, kaldi::nnet3::NnetComputation*)
kaldi::nnet3::CachingOptimizingCompiler::CompileNoShortcut(kaldi::nnet3::ComputationRequest const&)
kaldi::nnet3::CachingOptimizingCompiler::CompileAndCache(kaldi::nnet3::ComputationRequest const&)
kaldi::nnet3::CachingOptimizingCompiler::CompileInternal(kaldi::nnet3::ComputationRequest const&)
kaldi::nnet3::CachingOptimizingCompiler::CompileViaShortcut(kaldi::nnet3::ComputationRequest const&)
kaldi::nnet3::CachingOptimizingCompiler::CompileAndCache(kaldi::nnet3::ComputationRequest const&)
kaldi::nnet3::CachingOptimizingCompiler::CompileInternal(kaldi::nnet3::ComputationRequest const&)
kaldi::nnet3::CachingOptimizingCompiler::Compile(kaldi::nnet3::ComputationRequest const&)
kaldi::nnet3::NnetChainTrainer::Train(kaldi::nnet3::NnetChainExample const&)
main
__libc_start_main
nnet3-chain-train() [0xcdad99]
```

**Comment by danpovey** [comment]
OK, I have just pushed a code change that should resolve that-- please see
if it's better.

**Comment by GaofengCheng** [comment]
@danpovey it seems conv layer learn much faster than tdnn and lstm, about 5* faster, using the `learning-rate-factor=0.333 max-change=0.25` can halve the learning speed, but still too high.  And from several cnn experiments, cnn+batchnorm+(lstm)+tdnn overfit severely , from iter58 to 88, it becomes worse, using small change can relieve, but it still overfit , maybe I can try less epoch and dropout? .see:
cnn+tdnn:
`exp/ihm/chain_cleaned/tdnn1e_cnn_64dim_sp_bi: num-iters=89 nj=2..12 num-params=6.0M dim=64+100->3765 combine=-1.078->-0.283 xent:train/valid[58,88,final]=(-2.91,-2.33,-3.05/-2.98,-2.44,-3.13) logprob:train/valid[58,88,final]=(-0.210,-0.139,-0.253/-0.230,-0.165,-0.280)`

`exp/ihm/chain_cleaned/tdnn1e_cnn_64dim_low_maxchange_sp_bi: num-iters=89 nj=2..12 num-params=6.0M dim=64+100->3765 combine=-0.612->-0.231 xent:train/valid[58,88,final]=(-2.60,-2.43,-2.78/-2.66,-2.51,-2.85) logprob:train/valid[58,88,final]=(-0.172,-0.155,-0.205/-0.187,-0.176,-0.225)`

cnn+tdnn+lstm:
`exp/ihm/chain_cleaned/tdnn1e_cnn_latm_batchnorm_64dim_sp_bi/: num-iters=89 nj=2..12 num-params=46.6M dim=64+100->3765 combine=-0.308->-0.105 xent:train/valid[58,88,final]=(-1.16,-0.831,-1.12/-1.33,-1.14,-1.34) logprob:train/valid[58,88,final]=(-0.094,-0.057,-0.087/-0.121,-0.108,-0.124)`

`exp/ihm/chain_cleaned/tdnn1e_cnn_latm_batchnorm_64dim_small_maxchange_sp_bi: num-iters=89 nj=2..12 num-params=46.6M dim=64+100->3765 combine=-0.199->-0.102 xent:train/valid[58,88,final]=(-1.15,-0.887,-1.09/-1.32,-1.17,-1.33) logprob:train/valid[58,88,final]=(-0.087,-0.061,-0.083/-0.120,-0.112,-0.129)`

**Comment by GaofengCheng** [comment]
there is a mistake in previous comments, delete

**Comment by danpovey** [comment]
What was the mistake?
I notice that your 'final' probs were much worse than the other probs,
indicating that something went wrong in the combination stage.
BTW, re-running the compute_prob_(train or valid).sh with --test-mode=false
will indicate if the problem is unsuitable batch-norm stats.    If the
batch-norm stats are unsuitable, it will affect decoding.

Dan

**Comment by GaofengCheng** [comment]
@danpovey OK, after several CNN experiments, I think there may be problems in batchnorm and combination.
I do experiments: 
`batchcnn+batchtdnn VS batchcnn(smallmax-change)+batchtdnn VS cnn+batchtdnn VS cnn(smallmax-change)+batchtdnn`
also
`batchcnn+batchtdnn+lstm VS batchcnn(smallmax-change)+batchtdnn+lstm VS cnn+batchtdnn+lstm VS cnn(smallmax-change)+batchtdnn+lstm`
take cnn+tdnn+lstm for example (AMI IHM):
![image](https://cloud.githubusercontent.com/assets/7532101/25464491/40de8342-2b2f-11e7-8771-a67185d61fe9.png)

Anyway, I will follow your idea, try test-mode false

**Comment by danpovey** [comment]
It's more complicated than just doing test-mode=false, there needs to be a
command that refreshes the stats on the final.mdl before decoding.

Like one final iteration of training, but with zero learning rate, done
after combination.  That would do it.  I'll try to make a PR.

**Comment by GaofengCheng** [comment]
After your PR, will  `test-mode=false` still be needed?

**Comment by danpovey** [comment]
test-mode=false will mostly just affect the diagnostics, I'm not sure
whether I'll recommend to pass in that option; anyway we can change the
default at the C++ level if needed.

**Comment by GaofengCheng** [comment]
@danpovey I cannot understand , why model without batchnorm in cnn will perform such bad compared with their counterparts with batch in cnn, the accuracy for training iter looks good, though vibrating a lot(batchnorm in cnn make the training more stable). According to our guess, batchnorm stats lead to decoding problem. removing it in cnn should relieve this problem

**Comment by danpovey** [comment]
I don't really understand what you are asking.

**Comment by danpovey** [comment]
Gaofeng, within half an hour I will submit a PR.  But I'm only doing the
change for the nnet3, non-chain side, I will want you to replicate my
changes in the 'chain' binaries before testing it.  You can make a PR with
your changes.

**Comment by danpovey** [comment]
So what I do in my PR is, in the nnet3-combine code we use test-mode=false
to get the combination weights; and when we have them, we get the batchnorm
stats again, on the combination data.
I believe all the lines of code that I am changing have 'chain' versions
that need to be changed also.

**Comment by GaofengCheng** [comment]
@danpovey my doubt is why I just remove batch in cnn, the WER will change so much.....

**Comment by danpovey** [comment]
So maybe what you're saying that if you just remove batch-norm from CNN,
the WER gets bad.
I think CNNs tend to get very unstable without batchnorm, maybe you'd need
much smaller learning rates.
